File: 프로그래머스/lv3/42861. 섬 연결하기/섬 연결하기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

vector<int> board;

bool cmp(vector<int> &lhs, vector<int> &rhs) {
    return lhs[2] < rhs[2];
}

int find(int x) {
    if(board[x] == x) return x;
    else {
        return board[x] = find(board[x]);
    }
}

void unify(int x, int y) {
    x = find(x);
    y = find(y);
    if(x < y) board[y] = x;
    else board[x] = y;
}

int solution(int n, vector<vector<int>> costs) {
    int answer = 0;
    for(int i = 0; i < n; i++) board.push_back(i);
    sort(costs.begin(), costs.end(), cmp);
    
    for(vector<int> cost: costs) {
        if(find(cost[0]) == find(cost[1])) continue;
        unify(cost[0], cost[1]);
        answer += cost[2];
    }
    
    return answer;
}

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.py
parents = []
def Find(x):
    global parents
    if parents[x] == x:
        return x
    else:
        p = Find(parents[x])
        parents[x] = p
        return p

def Union(x, y):
    global parents
    x = Find(x)
    y = Find(y)
    
    if x < y:
        parents[y] = x
    else:
        parents[x] = y

def solution(n, costs):
    global parents
    parents = list(range(n))
    
    answer = 0
    
    costs.sort(key=lambda x: x[2])
    
    for x, y, cost in costs:
        if Find(x) == Find(y):
            continue
        else:
            Union(x, y)
            answer += cost
    
    return answer